src/query/shorthand.ts

```diff
@@ -277,3 +282,205 @@ export function fieldDef(encQ: EncodingQuery,
+        if (partParams[nestedPropertyParentIndex + nestedPropertyParent.length + 1] === '{') {
+          let openingBraceIndex = nestedPropertyParentIndex + nestedPropertyParent.length + 1;
+          closingBraceIndex = getClosingBraceIndex(openingBraceIndex, partParams);
+          encQ[nestedPropertyParent] = JSON.parse(partParams.substring(openingBraceIndex, closingBraceIndex + 1));
+
+        } else {
+          // Substring until the next comma (or end of the string)
+          let nextCommaIndex = partParams.indexOf(',', nestedPropertyParentIndex + nestedPropertyParent.length);
+          if (nextCommaIndex === -1) {
+            nextCommaIndex = partParams.length;
+          }
+
+          let parsedValue = JSON.parse(
+            partParams.substring(
+              nestedPropertyParentIndex + nestedPropertyParent.length + 1,
+              nextCommaIndex
+            )
+          );
+
+          if (nestedPropertyParent === 'maxbins') {
+            encQ.bin['maxbins'] = parsedValue;
+          } else {
+            encQ[nestedPropertyParent] = parsedValue;
+          }
+        }
+      }
+    }
+
+    return encQ;
+  }
+
+  // TODO(https://github.com/uwdata/compassql/issues/259):
+  // Extend this to support nested braces and brackets
+  export function getClosingBraceIndex(openingBraceIndex: number, str: string): number {
+    for (let i = openingBraceIndex; i < str.length; i++) {
+      if (str[i] === '}') {
+        return i;
+      }
+    }
+  }
+
+  export function fn(encQ: EncodingQuery, fieldDefShorthand: string): EncodingQuery {
+    // Aggregate, Bin, TimeUnit as enum spec case
+    if (fieldDefShorthand[0] === '?') {
+      let closingBraceIndex = getClosingBraceIndex(1, fieldDefShorthand);
+
+      let fnEnumIndex = JSON.parse(fieldDefShorthand.substring(1, closingBraceIndex + 1));
+
+      for (let encodingProperty in fnEnumIndex) {
+        if (isArray(fnEnumIndex[encodingProperty])) {
+          encQ[encodingProperty] = {enum: fnEnumIndex[encodingProperty]};
+        } else {
+          encQ[encodingProperty] = fnEnumIndex[encodingProperty];
+        }
+      }
+
+      return rawFieldDef(encQ,
+        splitWithTail(fieldDefShorthand.substring(closingBraceIndex + 2, fieldDefShorthand.length), ',', 2)
+      );
+    } else {
+      let func = fieldDefShorthand.substring(0, fieldDefShorthand.indexOf('('));
+      let insideFn = fieldDefShorthand.substring(func.length + 1, fieldDefShorthand.length - 1);
+      let insideFnParts = splitWithTail(insideFn, ',', 2);
+
+      if (AGGREGATE_OP_INDEX[func]) {
+        return aggregate(encQ, func, insideFnParts);
+      } else if (MULTI_TIMEUNIT_INDEX[func] || SINGLE_TIMEUNIT_INDEX[func]) {
+        return timeUnit(encQ, func, insideFnParts);
+      } else if (func === 'bin') {
+        return bin(encQ, func, insideFnParts);
+      }
+    }
+  }
+
+  export function aggregate(encQ: EncodingQuery, aggregate: string, insideFnParts: string[]): EncodingQuery {
```

Comment:
aggregate, timeUnit, bin  all seem unnecessary.  They are all just one liners that do not worth extraction.  Why don't you just integrate them into `fn` method?  I don't why you would ever want to call these method separately.